### PR TITLE
[Linter] Checks for explicit self-assignments.

### PR DESCRIPTION
**Pull request by tx-tomcat**
## Description 
This code implements a linter check to detect self-assignments in Move code.
The check focuses on two types of expressions:
   1. Mutate expressions (`*a = b`)
   2. Assign expressions (`a = b`)
 
In each case, the compiler tracks for the same "memory locations" of the assignment, attempting to track the same local variable. It does not have any sort of aliasing notion nor does it track references in any meaningful way. 

## Test plan
Added more use case

## Release notes

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [X] CLI: Move lint now warns against unnecessary self-assignments. 
- [ ] Rust SDK:

### CODE REVIEW

File: external-crates/move/crates/move-compiler/tests/linter/self_assignment.move [hunk 1]
```diff
@@ -0,0 +1,16 @@
+module 0x42::M {
+
+    fun func1() {
+        let x = 5;
+        let y = 10;
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/tests/linter/self_assignment.move:5]: Could this be extended to handle, e.g., `let (x, y) = (x, y)` ?

File: external-crates/move/crates/move-compiler/src/linters/self_assignment.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                    LValue_::Var {
+                        var: sp!(_, lhs),
+                        ..
+                    }
+                ),
+            ) = value_list.value.get(0)
+            {
+                if let UnannotatedExp_::Copy {
+                    var: sp!(_, rhs), ..
+                } = &assign_exp.exp.value
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/linters/self_assignment.rs:62]: Should this also handle the `Move` case?

File: external-crates/move/crates/move-compiler/tests/linter/self_assignment.move (hunk 1 above)
Comment (jamecornel) [reply external-crates/move/crates/move-compiler/tests/linter/self_assignment.move]: `let (x, y) = (x, y)` does not work in sui move. Tuple only appear in expressions (usually in the return position for a function)

File: external-crates/move/crates/move-compiler/src/linters/self_assignment.rs (hunk 2 above)
Comment (jamecornel) [reply external-crates/move/crates/move-compiler/src/linters/self_assignment.rs:63]: refactored

File: external-crates/move/crates/move-compiler/src/linters/self_assignment.rs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            false
+        }
+    }
+
+    fn report_self_assignment(&mut self, loc: Loc) {
+        let msg = "Explicit self-assignment detected for variable. Consider removing it to clarify intent.".to_string();
+        let diag = diag!(SELF_ASSIGNMENT_DIAG, (loc, msg));
+        self.env.add_diag(diag);
+    }
+}
```
Comment (tx-tomcat) [inline external-crates/move/crates/move-compiler/src/linters/self_assignment.rs:145]: The function takes three parameters:
`&mut self:`A mutable reference to the struct this method belongs to.
`value_list:`A reference to an LValueList, which likely contains the left-hand side of the assignment.
`assign_exp`: A reference to the right-hand side of the assignment.
`loc`: A Loc value, probably representing the location of this assignment in the source code.

It first checks if the `assign_exp` is an `ExpList` (a list of expressions).
If it is, it then checks for self-assignment by iterating over the left-hand side values (`value_list.value`) and the right-hand side expressions (`rhs_expressions`) simultaneously.
For each pair, it checks if:

The left-hand side is a `LValue_::Var`.
The right-hand side is a `Single` expression.
The right-hand side expression is either a `Copy` or `Move` operation.
The variable being copied or moved on the right side is the same as the variable on the left side.

If all pairs in the assignment satisfy these conditions, `is_self_assignment` is set to true.

**Pull request merged**